### PR TITLE
Check for duplicate scopes

### LINKED ISSUE #23587: [Core] Add ability for the DiagnosticScopeValidatingInterceptor to check for duplicate Activities for the same method
This check would help catch issues like this - https://github.com/Azure/azure-sdk-for-net/issues/23545

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/23587

### CODE REVIEW

**Comment by seanmcc-msft** [comment]
I think we may need a little more work on the duplicate Diagnostic Scope check.

https://github.com/seanmcc-msft/azure-sdk-for-net/tree/feature/storage%2FduplicateScope

In this branch, I attempted to resolve all the Storage-related duplicate Scope instances.  Most were trivial, but it appears this check is failing for parallel, multi-part uploads.

The following tests are failing:

- DataLake UploadAsync_NonZeroStreamPositionMultipleBlocks()
- Blobs UploadAsync_NonZeroStreamPositionMultipleBlocks()
- Blobs UploadAsync_AccessConditions(bool multiPart)
- Blobs UploadAsync_NonZeroStreamPositionMultipleBlocks()
- Blobs Upload_ImmutableStorageWithVersioning(bool multipart)

I spoke with @jaschrep-msft, and our theory is that the duplicate scope check is unhappy we are staging multiple blocks in parallel in the same API.

**Comment by JoshLove-msft** [comment]
> I think we may need a little more work on the duplicate Diagnostic Scope check.
> 
> https://github.com/seanmcc-msft/azure-sdk-for-net/tree/feature/storage%2FduplicateScope
> 
> In this branch, I attempted to resolve all the Storage-related duplicate Scope instances. Most were trivial, but it appears this check is failing for parallel, multi-part uploads.
> 
> The following tests are failing:
> 
> * DataLake UploadAsync_NonZeroStreamPositionMultipleBlocks()
> * Blobs UploadAsync_NonZeroStreamPositionMultipleBlocks()
> * Blobs UploadAsync_AccessConditions(bool multiPart)
> * Blobs UploadAsync_NonZeroStreamPositionMultipleBlocks()
> * Blobs Upload_ImmutableStorageWithVersioning(bool multipart)
> 
> I spoke with @jaschrep-msft, and our theory is that the duplicate scope check is unhappy we are staging multiple blocks in parallel in the same API.

Ah, I see. Yes, the check currently looks at all active scopes so running the same method in parallel would cause a false-positive. I will see what we can do to fix this.

File: sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs [hunk 1]
```diff
@@ -141,6 +146,21 @@ public void Dispose()
 
             foreach (ProducedDiagnosticScope producedDiagnosticScope in Scopes)
             {
+                var activity = producedDiagnosticScope.Activity;
+                var operationName = activity.OperationName;
+                // traverse the activities and check for duplicates among ancestors
+                while (activity != null)
+                {
+                    if (operationName == activity.Parent?.OperationName)
```
Comment (jsquire) [inline sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs:154]: Any chance `operationName` could be null here and cause a false positive?

**Review by jsquire — APPROVED** [review]

**Comment by seanmcc-msft** [comment]
@JoshLove-msft, all Storage tests are now passing, I have pushed my commits to this branch.

File: sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs:154]: I don't think so, but if it was null there is something wrong so it wouldn't be a false-positive 😄

File: sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs [hunk 2]
```diff
@@ -141,6 +146,21 @@ public void Dispose()
...
+                    {
+                        // Throw this exception lazily on Dispose, rather than when the scope is started, so that we don't trigger a bunch of other
+                        // erroneous exceptions relating to scopes not being completed/started that hide the actual issue
+                        throw new InvalidOperationException($"A scope has already started for event '{producedDiagnosticScope.Name}'");
+                    }
+
+                    _activityGraph.TryGetValue(activity, out activity);
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs:161]: Why don't you do `activity = activity.Parent`?

**Review by pakrym — APPROVED** [review]

File: sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs:161]: Not sure what I was thinking..

**Comment by JoshLove-msft** [comment]
@drwill-ms - can you help out with the DigitalTwin test failures?

**Review by seanmcc-msft — APPROVED** [review]

**Comment by drwill-ms** [comment]
> @drwill-ms - can you help out with the DigitalTwin test failures?

@abhipsaMisra is looking at it for our team. The changes we could make are probably super simple, but we'd like some guidance on which option we should take.

**Comment by JoshLove-msft** [comment]
> > @drwill-ms - can you help out with the DigitalTwin test failures?
> 
> @abhipsaMisra is looking at it for our team. The changes we could make are probably super simple, but we'd like some guidance on which option we should take.

Sure, feel free to ping me with any questions.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs [hunk 3]
```diff
@@ -21,37 +21,27 @@ internal partial class DigitalTwinModelsRestClient
...
-                            }
-                            value = array;
-                            return Response.FromValue(value, message.Response);
+                            array.Add(DigitalTwinsModelData.DeserializeDigitalTwinsModelData(item));
                         }
-                    default:
-                        throw await _clientDiagnostics.CreateRequestFailedExceptionAsync(message.Response).ConfigureAwait(false);
-                }
-            }
-            catch (Exception ex)
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs:51]: here and below - we don't need to catch and rethrow the exception anymore. The details are captured in the diagnostic scope in the convenience layer.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs [hunk 4]
```diff
@@ -21,37 +21,27 @@ internal partial class DigitalTwinModelsRestClient
             CreateModelsOptions digitalTwinModelsAddOptions = null,
             CancellationToken cancellationToken = default)
         {
-            using DiagnosticScope scope = _clientDiagnostics.CreateScope("DigitalTwinModelsClient.Add");
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs:24]: here and below - all of these protocol layer calls have corresponding convenience layer calls in which the diagnostic scope is created and started.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/DigitalTwinModelsRestClient.cs:51]: Thanks @abhipsaMisra! Will merge this in once the CI turns green.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**